ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionCommitterSpec.scala

```diff
@@ -595,58 +595,125 @@ class TransactionCommitterSpec extends AnyWordSpec with Matchers with MockitoSug
+    ): SubmittedTransaction = {
+      val lookup = txBuilder.lookupByKey(createNode(contractId = s"#$freshContractId"), found)
+      val builder = TransactionBuilder()
+      if (inRollback) {
+        val rollback = builder.add(txBuilder.rollback())
+        builder.add(lookup, rollback)
+      } else {
+        builder.add(lookup)
+      }
+      builder.buildSubmitted()
+    }
+
+    val mykey = {
+      val aCreateNode = createNode("#dummy")
+      Conversions.encodeContractKey(aCreateNode.templateId, aCreateNode.key.get.key)
+    }
 
-      val lookupByKeyNotFound =
-        newLookupByKeySubmittedTransaction(found = false)
-      val lookupByKeyFound1 = newLookupByKeySubmittedTransaction(found = true)
-      val lookupByKeyFound2 = newLookupByKeySubmittedTransaction(found = true)
+    "return Inconsistent when a contract key resolves to a different contract ID than submitted by a participant" in {
+
+      val cases =
+        Seq((false, Some(s"#$freshContractId")), (true, Some(s"#$freshContractId")), (true, None))
+          .flatMap { case (found, contractIdAtCommitter) =>
+            Seq(false, true).map(inRollback =>
+              newLookupByKeySubmittedTransaction(found, inRollback) -> contractIdAtCommitter
+            )
+          }
 
-      val cases = Table(
+      val casesTbl = Table(
         "transaction" -> "contractIdAtCommitter",
-        lookupByKeyNotFound -> Some(s"#$freshContractId"),
-        lookupByKeyFound1 -> Some(s"#$freshContractId"),
-        lookupByKeyFound2 -> None,
+        cases: _*
       )
 
-      val key = {
-        val aCreateNode = createNode("#dummy")
-        Conversions.encodeContractKey(aCreateNode.templateId, aCreateNode.key.get.key)
-      }
-      forAll(cases) { (transaction: SubmittedTransaction, contractIdAtCommitter: Option[String]) =>
-        val context = commitContextWithContractStateKeys(
-          key -> contractIdAtCommitter
-        )
-        val result =
-          ContractKeysValidation
-            .validateKeys(transactionCommitter)
-            .apply(
-              context,
-              DamlTransactionEntrySummary(createTransactionEntry(List("Alice"), transaction)),
-            )
-        result shouldBe a[StepStop]
-        val rejectionReason =
-          getTransactionRejectionReason(result).getInconsistent.getDetails
-        rejectionReason should startWith("InconsistentKeys")
+      forAll(casesTbl) {
+        (transaction: SubmittedTransaction, contractIdAtCommitter: Option[String]) =>
+          val context = commitContextWithContractStateKeys(
+            mykey -> contractIdAtCommitter
+          )
+          val result = validate(context, transaction)
+          result shouldBe a[StepStop]
+          val rejectionReason =
+            getTransactionRejectionReason(result).getInconsistent.getDetails
+          rejectionReason should startWith("InconsistentKeys")
       }
     }
 
+    "return DuplicateKeys when two local contracts conflict" in {
```

Comment:
Isn't this checked by the engine?

Reply:
No it’s not, I’m trying to get us there but atm the engine does absolutely no duplicate key checks and this allows you to create duplicate keys in some cases that kvutils does not catch.